### PR TITLE
Add subcommand etcdctl for kwokctl

### LINKED ISSUE #117: [kwokctl] Add subcommand `kwokctl etcdctl`
### What would you like to be added?

Add  subcommand `kwokctl etcdctl` like `kwokctl kubectl`

### Why is this needed?

Easy to manage etcd in kwok

### PR DESCRIPTION
**Pull request by Fish-pro**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:

Fixes #117 

#### Does this PR introduce a user-facing change?

```release-note
Add subcommand etcdctl for kwokctl
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kwokctl/runtime/cluster.go [hunk 1]
```diff
@@ -207,6 +207,25 @@ func (c *Cluster) KubectlInCluster(ctx context.Context, stm utils.IOStreams, arg
...
+	if err != nil {
+		bin := utils.PathJoin(conf.Workdir, "bin")
+		etcdctlPath = utils.PathJoin(bin, "etcdctl"+vars.BinSuffix)
+		err = utils.DownloadWithCacheAndExtract(ctx, conf.CacheDir, conf.EtcdBinaryTar, etcdctlPath, "etcdctl"+vars.BinSuffix, 0755, conf.QuietPull, true)
+		if err != nil {
+			return err
+		}
+	}
+
+	return utils.Exec(ctx, "", stm, etcdctlPath, args...)
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/cluster.go:226]: For binary runtime, just specify the port of etcd, 
For kind runtime, may need to expose etcd's port to local with `kubectl port-forward` and use it
For docker runtime, also need to expose etcd's port to local, may use `docker exec -i ${etcdpodname} nc 127.0.0.1 ${etcdport}` can be work.

Comment (Fish-pro) [reply pkg/kwokctl/runtime/cluster.go]: It has been modified, thanks for your guidance. I have considered that kind is not supported at the beginning, so I removed the parameter. If kubectl port-forward is used, I can understand it no problem

Comment (wzshiming) [reply pkg/kwokctl/runtime/cluster.go]: Add a test, Try to get a value using `kwokctl etcdctl` in this

https://github.com/kubernetes-sigs/kwok/blob/96f9165f68a0d71c587b32f935397665638985fc/test/kwokctl/kwokctl_workable_test.sh#L49

Comment (wzshiming) [reply pkg/kwokctl/runtime/cluster.go]: By the way I have an idea, for `kind` and `docker` you can call the `etcdctl` directly from the container like `snapshot`.

Comment (wzshiming) [reply pkg/kwokctl/runtime/cluster.go]: https://github.com/kubernetes-sigs/kwok/blob/ca4b082cec3bda5083b5080cbaf74803266583e2/pkg/kwokctl/runtime/compose/cluster_snapshot.go

https://github.com/kubernetes-sigs/kwok/blob/a98d86be2828090a7127d2033b7407bea111a40e/pkg/kwokctl/runtime/kind/cluster_snapshot.go

Comment (wzshiming) [reply pkg/kwokctl/runtime/cluster.go]: EtcdctlInCluster should be implemented independently in each runtime's directory, as it is not universal to each other

Comment (wzshiming) [reply pkg/kwokctl/runtime/cluster.go]: @Fish-pro

Comment (Fish-pro) [reply pkg/kwokctl/runtime/cluster.go]: @wzshiming Has been implemented separately at each runtime, the test script has a path problem, can I submit it in another pr

**Comment by Fish-pro** [comment]
@wzshiming 
Here are the test results when runtime is kind
```sh
➜  kwok git:(feature/etcdctl) kwokctl create cluster --name kwok --runtime kind --kind-node-image kindest/node:v1.21.1
Creating cluster                                                    cluster=kwok
Starting cluster                                                    cluster=kwok
Creating cluster "kwok-kwok" ...
 ✓ Ensuring node image (kindest/node:v1.21.1) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kwok-kwok"
You can now use your cluster with:

kubectl cluster-info --context kind-kwok-kwok

Thanks for using kind! 😊
Cluster is ready                                                    cluster=kwok
You can now use your cluster with:

    kubectl config use-context kwok-kwok

Thanks for using kwok!
➜  kwok git:(feature/etcdctl) kwokctl create cluster --name kwok --runtime kind --kind-node-image kindest/node:v1.21.1
➜  kwok git:(feature/etcdctl) go run cmd/kwokctl/main.go etcdctl get /registry/namespaces/default --print-value-only

v1	Namespace�
�
default"*$489d09b9-c286-40a3-8833-121c69c0cdcc2ī��Z&

kube-apiserverUpdatevī��FieldsV1:I
G{"f:metadata":{"f:labels":{".":{},"f:kubernetes.io/metadata.name":{}}}}


kubernetes
Active"
```

File: pkg/kwokctl/runtime/compose/cluster.go [hunk 2]
```diff
@@ -423,3 +423,29 @@ func (c *Cluster) buildComposeCommands(ctx context.Context, args ...string) ([]s
...
+	}
+
+	bin := utils.PathJoin(conf.Workdir, "bin")
+	etcdctlPath := utils.PathJoin(bin, "etcdctl"+vars.BinSuffix)
+
+	err = utils.DownloadWithCacheAndExtract(ctx, conf.CacheDir, conf.EtcdBinaryTar, etcdctlPath, "etcdctl"+vars.BinSuffix, 0755, conf.QuietPull, true)
+	if err != nil {
+		return err
+	}
+
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/compose/cluster.go:441]:
```suggestion
```

File: pkg/kwokctl/runtime/kind/cluster.go [hunk 3]
```diff
@@ -382,3 +382,34 @@ func (c *Cluster) ListImages(ctx context.Context, actual bool) ([]string, error)
...
+	}
+
+	bin := utils.PathJoin(conf.Workdir, "bin")
+	etcdctlPath := utils.PathJoin(bin, "etcdctl"+vars.BinSuffix)
+
+	err = utils.DownloadWithCacheAndExtract(ctx, conf.CacheDir, conf.EtcdBinaryTar, etcdctlPath, "etcdctl"+vars.BinSuffix, 0755, conf.QuietPull, true)
+	if err != nil {
+		return err
+	}
+
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/kind/cluster.go:400]:
```suggestion
```

File: pkg/kwokctl/runtime/compose/cluster.go [hunk 4]
```diff
@@ -423,3 +423,29 @@ func (c *Cluster) buildComposeCommands(ctx context.Context, args ...string) ([]s
...
+	etcdContainerName := conf.Name + "-etcd"
+
+	// Save to /snapshot.db on container
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/compose/cluster.go:444]:
```suggestion
```

**Review by wzshiming — CHANGES_REQUESTED** [review]

**Comment by wzshiming** [comment]
> @wzshiming Here are the test results when runtime is kind
> 
> ```shell
> ➜  kwok git:(feature/etcdctl) kwokctl create cluster --name kwok --runtime kind --kind-node-image kindest/node:v1.21.1
> Creating cluster                                                    cluster=kwok
> Starting cluster                                                    cluster=kwok
> Creating cluster "kwok-kwok" ...
>  ✓ Ensuring node image (kindest/node:v1.21.1) 🖼
>  ✓ Preparing nodes 📦
>  ✓ Writing configuration 📜
>  ✓ Starting control-plane 🕹️
>  ✓ Installing CNI 🔌
>  ✓ Installing StorageClass 💾
> Set kubectl context to "kind-kwok-kwok"
> You can now use your cluster with:
> 
> kubectl cluster-info --context kind-kwok-kwok
> 
> Thanks for using kind! 😊
> Cluster is ready                                                    cluster=kwok
> You can now use your cluster with:
> 
>     kubectl config use-context kwok-kwok
> 
> Thanks for using kwok!
> 
> ➜  kwok git:(feature/etcdctl) go run cmd/kwokctl/main.go etcdctl get /registry/namespaces/default --print-value-only
> 
> v1	Namespace�
> �
> default"*$489d09b9-c286-40a3-8833-121c69c0cdcc2ī��Z&
> 
> kube-apiserverUpdatevī��FieldsV1:I
> G{"f:metadata":{"f:labels":{".":{},"f:kubernetes.io/metadata.name":{}}}}
> 
> 
> kubernetes
> Active"
> ```


Great, now add that test directly to e2e to keep that feature working all the time.

**Comment by wzshiming** [comment]
https://github.com/kubernetes-sigs/kwok/blob/b41042c16b1cf180cc8147fddf8d8c053f6b0aa5/pkg/kwokctl/runtime/compose/compose.yaml.tpl#L52

https://github.com/kubernetes-sigs/kwok/blob/e3724a6a88be6a60c659843ae239ae48d4896240/pkg/kwokctl/runtime/binary/cluster.go#L224

Let the etcd prefix in both runtimes follow the Kind runtime prefix `/registry`

**Comment by Fish-pro** [comment]
> runtimes

@wzshiming Thanks for your guidance, it has been revised

**Comment by wzshiming** [comment]
Add a test, Get a key of `/registry/namespaces/default` using `kwokctl etcdctl` in this

https://github.com/kubernetes-sigs/kwok/blob/96f9165f68a0d71c587b32f935397665638985fc/test/kwokctl/kwokctl_workable_test.sh#L49

File: pkg/kwokctl/runtime/binary/cluster.go [hunk 5]
```diff
@@ -676,3 +676,25 @@ func (c *Cluster) ListImages(ctx context.Context, actual bool) ([]string, error)
...
+	}
+	bin := utils.PathJoin(conf.Workdir, "bin")
+	etcdctlPath := utils.PathJoin(bin, "etcdctl"+vars.BinSuffix)
+
+	err = utils.DownloadWithCacheAndExtract(ctx, conf.CacheDir, conf.EtcdBinaryTar, etcdctlPath, "etcdctl"+vars.BinSuffix, 0755, conf.QuietPull, true)
+	if err != nil {
+		return err
+	}
+
+	utils.Exec(ctx, "", stm, etcdctlPath, append([]string{"--endpoints", "127.0.0.1:" + utils.StringUint32(conf.EtcdPort)}, args...)...)
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/binary/cluster.go:694]:
```suggestion
	err = utils.Exec(ctx, "", stm, etcdctlPath, append([]string{"--endpoints", "127.0.0.1:" + utils.StringUint32(conf.EtcdPort)}, args...)...)
```

**Review by wzshiming — CHANGES_REQUESTED** [review]

File: pkg/kwokctl/runtime/binary/cluster.go (hunk 5 above)
Comment (Fish-pro) [reply pkg/kwokctl/runtime/binary/cluster.go]: @wzshiming Thanks

File: pkg/kwokctl/cmd/etcdctl/etcdctl.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	"sigs.k8s.io/kwok/pkg/kwokctl/utils"
+	"sigs.k8s.io/kwok/pkg/kwokctl/vars"
+	"sigs.k8s.io/kwok/pkg/log"
+)
+
+type flagpole struct {
+	Name string
+}
+
+// NewCommand returns a new cobra.Command for getting the list of clusters
```
Comment (wzshiming) [inline pkg/kwokctl/cmd/etcdctl/etcdctl.go:36]:
```suggestion
// NewCommand returns a new cobra.Command for use kwokctl in a cluster
```

**Review by wzshiming — CHANGES_REQUESTED** [review]
Modify the last comment, the other LGTM

**Comment by Fish-pro** [comment]
> other

It has been modified

**Pull request merged**